File: solutions/Java/WordSearchBFS.java
import java.util.*;

class Solution {
    public boolean exist(char[][] board, String word) {
        if (board == null || board.length == 0 || word == null || word.length() == 0) {
            return false;
        }

        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] == word.charAt(0) && bfs(board, word, i, j)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean bfs(char[][] board, String word, int startRow, int startCol) {
        Queue<int[]> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        queue.offer(new int[]{startRow, startCol, 0});
        visited.add(startRow + "," + startCol);

        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int row = current[0], col = current[1], index = current[2];

            if (index == word.length() - 1) {
                return true;
            }

            for (int[] dir : directions) {
                int newRow = row + dir[0];
                int newCol = col + dir[1];
                String state = newRow + "," + newCol;

                if (newRow >= 0 && newRow < board.length &&
                    newCol >= 0 && newCol < board[0].length &&
                    board[newRow][newCol] == word.charAt(index + 1) &&
                    !visited.contains(state)) {

                    visited.add(state);
                    queue.offer(new int[]{newRow, newCol, index + 1});
                }
            }
        }

        return false;
    }
}

File: solutions/C++/word_search_bfs.cpp
#include <vector>
#include <string>
#include <queue>
#include <set>
using namespace std;

bool exist(vector<vector<char>>& board, string word) {
    for (int i = 0; i < board.size(); i++) {
        for (int j = 0; j < board[0].size(); j++) {
            if (board[i][j] == word[0]) {
                queue<tuple<int, int, int, set<pair<int, int>>>> q;
                set<pair<int, int>> visited;
                visited.insert({i, j});
                q.push({i, j, 1, visited});

                while (!q.empty()) {
                    auto [row, col, idx, vis] = q.front();
                    q.pop();
                    if (idx == word.length()) return true;

                    int dirs[4][2] = {{0,1}, {0,-1}, {1,0}, {-1,0}};
                    for (auto& d : dirs) {
                        int nr = row + d[0], nc = col + d[1];
                        if (nr >= 0 && nr < board.size() && nc >= 0 && nc < board[0].size() &&
                            board[nr][nc] == word[idx] && vis.find({nr, nc}) == vis.end()) {
                            auto newVis = vis;
                            newVis.insert({nr, nc});
                            q.push({nr, nc, idx + 1, newVis});
                        }
                    }
                }
            }
        }
    }
    return false;
}
